Fix quantifiers with nested statement-expressions

qinheping opened the pull request:
- Added support for `ID_address_of` and 'ID_typecast` expressions to `find_base_symbol`.
- Corrected the multi-path handling in `convert_statement_expression`---a path not reaching the terminal expression should result in `false`.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on regression/cbmc/Quantifiers-statement-expression3/main.c:
```diff
@@ -0,0 +1,13 @@
+  // clang-format off
+  // no side effects!
+  int j;
+  //assert(j++);
+  //assert(({int i = 0; while(i <3) i++; i <3;}));
```
What to make of these commented-out lines?

qinheping replied:
They are leftover. I will remove them.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -151,9 +170,13 @@ static exprt convert_statement_expression(
+
+    if(current_it == where.instructions.end())
+    {
+      res = and_exprt(res, implies_exprt(path_condition, true_exprt()));
```
Why are you constructing the expression in exactly this way? `implies_exprt(path_condition, true_exprt())` is semantically equivalent to `true_exprt()` after all?!

qinheping replied:
Yes, you are right. I will fix it.

qinheping committed:
Fix quantifiers with nested statement-expressions

- Added support for `ID_address_of` and 'ID_typecast` expressions to `find_base_symbol`.
- Corrected the multi-path handling in
  `convert_statement_expression`---a path not reaching the terminal
expression should be ignored.
